source/common/http/http2/codec_impl.cc

```diff
@@ -1016,25 +1036,43 @@ void ConnectionImpl::sendSettings(
   }
 }
 
+bool ConnectionImpl::setAndCheckNghttp2CallbackStatus(Status&& status) {
+  nghttp2_callback_status_ = std::move(status);
+  return nghttp2_callback_status_.ok();
+}
+
 ConnectionImpl::Http2Callbacks::Http2Callbacks() {
   nghttp2_session_callbacks_new(&callbacks_);
   nghttp2_session_callbacks_set_send_callback(
       callbacks_,
       [](nghttp2_session*, const uint8_t* data, size_t length, int, void* user_data) -> ssize_t {
-        return static_cast<ConnectionImpl*>(user_data)->onSend(data, length);
+        auto status_or_len = static_cast<ConnectionImpl*>(user_data)->onSend(data, length);
+        if (status_or_len.ok()) {
+          return status_or_len.value();
+        }
+        auto status = status_or_len.status();
+        static_cast<ConnectionImpl*>(user_data)->setAndCheckNghttp2CallbackStatus(
+            std::move(status));
+        return NGHTTP2_ERR_CALLBACK_FAILURE;
       });
 
   nghttp2_session_callbacks_set_send_data_callback(
       callbacks_,
       [](nghttp2_session*, nghttp2_frame* frame, const uint8_t* framehd, size_t length,
          nghttp2_data_source* source, void*) -> int {
         ASSERT(frame->data.padlen == 0);
-        return static_cast<StreamImpl*>(source->ptr)->onDataSourceSend(framehd, length);
+        auto status = static_cast<StreamImpl*>(source->ptr)->onDataSourceSend(framehd, length);
+        bool ok = static_cast<StreamImpl*>(source->ptr)
+                      ->parent_.setAndCheckNghttp2CallbackStatus(std::move(status));
+        return ok ? 0 : NGHTTP2_ERR_CALLBACK_FAILURE;
```

Comment:
can this pattern be moved into a helper?

Reply:
done